packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx

```diff
@@ -2233,6 +2233,31 @@ describe('shallow', () => {
       expect(wrapper.state()).to.eql({ id: 'foo' });
       expect(() => wrapper.setState({ id: 'bar' }, 1)).to.throw(Error);
     });
+
+    it('should preserve the receiver', () => {
+      class Comp extends React.Component {
+        constructor(...args) {
+          super(...args);
+
+          this.state = {
+            key: '',
+          };
+
+          this.instanceFunction = () => this.setState(() => ({ key: 'value' }));
+        }
+
+        componentDidMount() {
+          this.instanceFunction();
+        }
+
+        render() {
+          const { key } = this.state;
+          return key ? null : null;
+        }
+      }
+
+      expect(shallow(<Comp />).debug()).to.equal('');
```

Comment:
Does `setState` in `componentDidMount` invoke `componentDidUpdate`?

Reply:
it should; cDU is invoked whenever it updates.

Reply:
In order to be called cDU, we should stub `setState` so I think we should move calling cDM after the stubbing.

Reply:
I'll write the tests and the patch.
https://github.com/airbnb/enzyme/compare/fix_shallow_setstate...koba04:fix_shallow_setstate?expand=1